Use home-grown RPC server instead of net/rpc.Server.

bdarnell opened the pull request:
The wire protocol is unchanged; the biggest difference in this commit is
that method registration is now explicit instead of being
reflection-based and the method signatures have been changed. This
change is motivated by a desire to replace the goroutine-spawning
behavior of net/rpc.Server (see #1669), although this commit does not
change that behavior yet.

bdarnell committed:
Use home-grown RPC server instead of net/rpc.Server.

The wire protocol is unchanged; the biggest difference in this commit is
that method registration is now explicit instead of being
reflection-based and the method signatures have been changed. This
change is motivated by a desire to replace the goroutine-spawning
behavior of net/rpc.Server (see #1669), although this commit does not
change that behavior yet.

MBerhault commented on rpc/server.go:
```diff
@@ -252,16 +312,84 @@ func (s *Server) Close() {
+	}()
+
+	for {
+		req, method, args, err := s.readRequest(codec)
```
method shadows method L36 (from make check)

tamird commented on rpc/send_test.go:
```diff
@@ -201,11 +200,12 @@ func TestClientNotReady(t *testing.T) {
+		methods: map[string]method{},
 	}
-	if err := s.Register(&Heartbeat{}); err != nil {
+	heartbeat := &Heartbeat{}
```
nit: doesn't look like you need this local (nor does it need to be a reference?)

bdarnell replied:
Yeah, it's not really necessary. I pulled it out because `(&Heartbeat{}).Ping` is ugly. I re-inlined it, although it does need to be a reference.

BramGruneir commented on rpc/server.go:
```diff
@@ -252,16 +312,84 @@ func (s *Server) Close() {
+// have finished), we close the responses channel.
+func (s *Server) readRequests(codec rpc.ServerCodec, responses chan<- serverResponse) {
+	var wg sync.WaitGroup
+	defer func() {
```
Why defer this and not just put it at the end of the function?

tamird replied:
because the loop contains `return`

BramGruneir replied:
cool

BramGruneir commented on rpc/server.go:
```diff
@@ -252,16 +312,84 @@ func (s *Server) Close() {
-		for _, cb := range s.closeCallbacks {
-			cb(conn)
+// readRequests synchronously reads a stream of requests from a
+// connection.  Each request is handled in a new background goroutine;
```
single spaces after periods... even though I prefer the double space too.

petermattis commented on rpc/server.go:
```diff
@@ -51,25 +61,51 @@ type Server struct {
+		return util.Errorf("method %s already registered", name)
+	}
+	reqType := reflect.TypeOf(reqPrototype)
+	if reqType.Kind() != reflect.Ptr {
```
Is the request object always a `proto.Message`? If so, might be better to test that `reqType` is convertible to `proto.Message`.

petermattis replied:
Or just change the `handler` signature to `func(proto.Message) (proto.Message, error)`.

tamird replied:
not true of gossip, at least

tamird replied:
er, i stand corrected

bdarnell replied:
For our codec, it is. This server could in principle work with `interface{}` if you gave it a different codec but we don't really care about that. I'll try changing all uses of `interface{}` to `proto.Message`.

bdarnell replied:
Changing `interface{}` to `proto.Message` in the handler signature results in import conflicts: any file that defines a handler needs to deal with our `proto` package; if we require `proto.Message` in handler signatures then they'll also need to import that `proto` package under an alias (`gogoproto`). This is pretty ugly, but probably worthwhile to avoid surprises with pointers versus non-pointers.

petermattis replied:
Now that you've changed the signature of `handler`, I don't think checking the type of `reqPrototype` is buying you anything. Remove it?

bdarnell replied:
In the current version of gogoproto, all implementers of `gogoproto.Message` are pointer types, but this is not necessarily guaranteed. We don't currently support hypothetical value types that implement the Message interface because we'd need to change the reflection logic in `readRequest` to do so. This is now an extra sanity check to guard against any introduction of value types that implement `gogoproto.Message`. 

It is theoretically possible for a value type to implement `gogoproto.Message` (although this doesn't happen in the current version of gogoproto). If this were to happen we would need to change the reflection code in `readRequest`.

spencerkimball commented on rpc/server.go:
```diff
@@ -51,25 +61,51 @@ type Server struct {
 
+// Register a new method handler. `name` is a qualified name of the form "Service.Name".
+// `handler` is a function that takes an argument of the same type as `reqPrototype`.
+// Both the argument and return value of 'handler' should be a pointer to a protocol
```
You don't actually verify it's a protocol message anywhere. Did you mean to?

bdarnell replied:
That verification happens down in the codec itself (although verifying it here will have much cleaner error handling).

petermattis commented on rpc/server.go:
```diff
@@ -252,16 +312,77 @@ func (s *Server) Close() {
+
+		wg.Add(1)
+		go func() {
+			reply, err := meth.handler(args)
```
`meth.handler`. Ha!

petermattis commented:
LGTM

bdarnell committed:
Define rpc interface in terms of proto.Message instead of interface{}

spencerkimball commented:
LGTM. Fast work.

tamird commented on rpc/server.go:
```diff
@@ -33,14 +34,24 @@ import (
 	"github.com/gogo/protobuf/proto"
```
i think we should call this import gogoproto everywhere

tamird commented:
LGTM

petermattis commented on rpc/server.go:
```diff
@@ -112,7 +149,31 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		s.sendResponses(codec, responses)
+		wg.Done()
+	}()
+	wg.Wait()
```
Not that it matters, but I think you could use the current goroutine for either reading requests or sending responses.

bdarnell replied:
Yeah. I originally opted for symmetry but I'll move readRequest into the current goroutine to save a few KB.

petermattis commented:
LGTM

bdarnell committed:
Remove an unnecessary goroutine.

bdarnell committed:
Fix error handling for unknown methods in rpc codec.

Previously, an unknown method would leave the connection in an
inconsistent state so the next request could not be parsed
correctly (the connection would probably error out but it would be
possible to parse the next frame as the wrong type of protobuf).

Added some tests for new rpc server.

bdarnell commented:
The last commit in this branch adds some tests and fixes a bug in the codec (@petermattis).

petermattis commented:
LGTM

tschottdorf commented on rpc/server.go:
```diff
@@ -33,14 +34,24 @@ import (
+}
+
 // Server is a Cockroach-specific RPC server with an embedded go RPC
 // server struct. By default it handles a simple heartbeat protocol
```
This comment is out of date.

tschottdorf commented on rpc/server.go:
```diff
@@ -252,16 +310,85 @@ func (s *Server) Close() {
+		wg.Add(1)
+		go func() {
+			reply, err := meth.handler(args)
+			responses <- serverResponse{
```
looks like this could still rearrange responses and that the changes here have only made it less likely (since the goroutine is spawned in a serial fashion). Am I missing something?

bdarnell replied:
This change just replaced the server implementation without changing the functionality; #1689 changed things so that raft messages are delivered in order.

tschottdorf commented on server/node.go:
```diff
@@ -472,82 +494,5 @@ func (n *nodeServer) executeCmd(args proto.Request, reply proto.Response) error
-	return nil
-}
-
-func (n *nodeServer) Get(args *proto.GetRequest, reply *proto.GetResponse) error {
```
cool we got rid of all those.

tschottdorf commented on server/raft_transport.go:
```diff
@@ -64,36 +65,35 @@ func newRPCTransport(gossip *gossip.Gossip, rpcServer *rpc.Server, rpcContext *r
-	if ok {
-		return server.RaftMessage(req, &multiraft.RaftMessageResponse{})
+	if !ok {
+		return nil, util.Errorf("Unable to proxy message to node: %d", req.Message.To)
```
s/Unable/unable/

tschottdorf commented:
LGTM

tschottdorf commented:
I'll bundle up the trivial comments.
